Add dashboard site documentation (#326)
hubverse-org/hubDocs: docs/source/developer/dashboard-site.md

# Generating the Dashboard Website

The [dashboard website
builder](https://github.com/hubverse-org/hub-dash-site-builder) is a [docker
image](#dashboard-tool-docker) which is thin wrapper around [the quarto
publishing system](#dashboard-tool-yq) and [yq](#dashboard-tool-yq). This tool
was designed under the following motivations:

1. Maintenance of this dashboard should be independent from hub maintenance.
2. Anyone should be able to deploy a dashboard as long as they have a hub with
   target data that they can access.
3. The only knowledge anyone should need is of Markdown, YAML, and basic
   operation of GitHub workflows. Knowledge of any particular [website
   generator](https://jamstack.org/generators/) is not required and the user
   should not need to handle any templating that cannot be easily automated.
4. We should not have to maintain a server to deploy these sites.

> ## References

Reviewer: I think this section would be better at the end of the topic

Author: done!

Here are links to some concepts that are used in the website builder.

- [Building a website using quarto](https://quarto.org/docs/websites/)
- [BASH scripting cheat sheet](https://devhints.io/bash)
- `yq` [tips and tricks](https://mikefarah.gitbook.io/yq/usage/tips-and-tricks)
  and [recipes](https://mikefarah.gitbook.io/yq/recipes). Specific operators
  can be found in the comments of the
  [`modify-quarto-yml.sh`](https://github.com/hubverse-org/hub-dash-site-builder/tree/main/modify-quarto-yml.sh)
  file in the hub-dash-site-builder repository.
- [Dockerfile reference](https://docs.docker.com/reference/dockerfile/)

## Installation

You can install the latest version of this tool with docker:

```bash
docker pull ghcr.io/hubverse-org/hub-dash-site-builder:latest
```

(dashboard-site-source)=
## How the website is built (a tale of two sources)

> 

Reviewer: I think it would be good to add a short section here called "Configuring the site" that lists what the user is able to configure. 

Then it will also be good to state below that the steps described below are effectively the auto-generated part of the site that takes the user configurations and builds the site from them.

The purpose of the suggestion is to make clear, what the workflow expects as input and is edited by users and what files are then used to auto-generate the site via the builder (which is perfectly described below!)

I know it's alluded to in the table below but I think a short separate section would help conceptually tie the information together better.

Author: Good point. I've added it in commit 41cf7ed9fd4e7d7eb873df70bf11aa8c77bdcedf

have the Forecast or Evals page. Instead, it has self-generated reports.

If a hub does not want to build either the forecast or evaluations pages, they
can omit the predevals or predtimechart config files. When this happens, the
site builder will remove these pages and the associated resources before
building the quarto site.

## How the visualizations work

If the dashboard is built independent from the data generation, how do the
visualizations work? They work because the visualization pages each call a
script that loads either predtimechart or predevals with a JavaScript function
that is designed to fetch data for these modules. All of the work of fetching
the data is done by the browser, but this does require data to exist.

For public hubs, these data live on a separate branch that we can access via
a `raw.githubusercontent.com` URL. For private hubs, data can be built locally
and bundled with the site.

## How the image is built

The images is built using GitHub actions and [deployed to GitHub's container registry](https://github.com/hubverse-org/hub-dash-site-builder/pkgs/container/hub-dash-site-builder/406574245?tag=latest).

> We initially cribbed the build workflow from [GitHub's Publishing Docker images

Reviewer: What does cribbed mean?

Author: ```suggestion
We initially created the build workflow by referencing [GitHub's Publishing Docker images
```

on the GitHub runner and saves it as a tar file. The image is then loaded and
tested against the reichlab/flusight-dashboard. At the end, assuming all tests
pass, the image is uploaded as an artifact.

### Test

- **purpose**: test the built docker image against the tests from the main branch.
- **permissions**: read-all
- **runs on**: pull request, and manual trigger not on main branch.

This job is needed to ensure the tests from the main branch continue to work and
are there to prevent potentially malicious pull requests from forcing tests to
pass. It has two steps:

1. fetch the image artifact
2. load and test the artifact against the tests as they exist on the main branch.

This is not run from the main branch or on a tag because by this time, the tests
from the build image job will be redundant.

### Publish

- **purpose**: publish the built docker image
- **permissions**: read-all
- **runs on**: push of a tag that starts with "v" and a workflow dispatch from
  main where "publish" is selected

This is similar to the build image except that instead of building the image,
we are loading it from an artifact and pushing it to the registry.

> The final step of this job is to generate an artifact attestation, which is a

Reviewer: what does attestation mean? Is it a technical term?

Author: This is a technical term and I think it kinda belongs here since this is developer documentation (not intended for general audiences).

Author: ```suggestion
The final step of this job is to generate [an artifact attestation](https://docs.github.com/en/actions/security-for-github-actions/using-artifact-attestations/using-artifact-attestations-to-establish-provenance-for-builds), which is a
```

# Generating the Dashboard Website

The [dashboard website
builder](https://github.com/hubverse-org/hub-dash-site-builder) is a [docker
image](#dashboard-tool-docker) which is thin wrapper around [the quarto
publishing system](#dashboard-tool-yq) and [yq](#dashboard-tool-yq). This tool
was designed under the following motivations:

1. Maintenance of this dashboard should be independent from hub maintenance.
2. Anyone should be able to deploy a dashboard as long as they have a hub with
   target data that they can access.
3. The only knowledge anyone should need is of Markdown, YAML, and basic
   operation of GitHub workflows. Knowledge of any particular [website
   generator](https://jamstack.org/generators/) is not required and the user
   should not need to handle any templating that cannot be easily automated.
4. We should not have to maintain a server to deploy these sites.

## Installation

You can install the latest version of this tool with docker:

```bash
docker pull ghcr.io/hubverse-org/hub-dash-site-builder:latest
```

(dashboard-site-source)=
## How the website is built (a tale of two sources)

> The website that is built is a fully static site that can be viewed locally. It

Reviewer 2: It would be helpful to link to the other docs page for the local dashboard workflow here

Author: I always get confused trying to reason about [myst-parser's implicit cross-links](https://myst-parser.readthedocs.io/en/latest/syntax/cross-referencing.html#inline-links-with-implicit-text), so this may or may not work, but I'll fix it when I render this locally. 

```suggestion
The website that is built is a fully static site that [can be viewed locally](./dashboard-local.md). It
```

| `about.html` | --- | `pages/about.md` | yes |
| `data.html` | --- | `pages/data.qmd` | yes |
| `img/` | --- | `pages/img` | yes |
| `eval.html` | `static/eval.qmd` | `predevals-config.yml` | yes |
| `forecast.html` | `static/eval.qmd` | `predtimechart-config.yml` | yes |
| `resources/` | `static/resources/` | --- | no |
| `site_libs` | `static/_quarto.yml` | --- | no |

### An incomplete boilerplate

On its own, the `static/` directory of the dashboard site builder contains an
incomplete quarto website. The incomplete parts are:

1. `index.qmd` is missing
2. The JavaScript in `resources/` is incomplete. There is a single line that
   defines the root folder for resource access, with a placeholder called
   `{ROOT}`, which needs to be replaced by a local folder or URL for the
   resource:
   ```js
   const root = "{ROOT}";
   ```

The site builder image effectively performs four steps to complete and render
the site:

1. Combine the contents of `pages/` and `/static/` into a temporary directory, `/tmp/`.
2. Use `yq` to merge `site-config.yml` into `/tmp/_quarto.yml`.
3. Update the JavaScript files to point to the correct resources or discard unused files.

> 4. Run `quarto render /tmp/` and copy the output to the output directory (`out/`).

Reviewer 2: Do we still have an `out/` directory? I thought it was replaced by `data/`

Author: There is no official directory name here. By default if you do not specify a directory name, it points to `pages/_site`. The reason why that's the default is because that was the behavior before version 1.0.0. The reason that was the behavior before version 1.0.0 is because I had initially written the script to copy the contents of `/site` into `pages` and running quarto on that folder without first creating a temporary folder. By default, quarto dumps the site into a folder called `_site/` at the root of the quarto project, so that ended up being `pages/_site`

Reviewer 2: Ok, can this information be incorporated somehow? I still find the original text a little confusing

Author: I've rearranged the section a bit and added a command line call to the container that hopefully helps.

Reviewer 2: Yes, it looks good now, thank you!

```{mermaid}
:config: {"theme": "base", "themeVariables": {"primaryColor": "#dbeefb", "primaryBorderColor": "#3c88be"}}
flowchart TD
  subgraph dashboard
    pages/contents["pages/[contents]"]
    site-config.yml
    out/
  end
  subgraph docker
    subgraph /static/
        /static/contents["/static/[contents]"]
        /static/_quarto.yml["/static/_quarto.yml"]
    end
    yq{{"yq"}}
    subgraph /tmp/
        _quarto.yml["/tmp/_quarto.yml"]
        contents["/tmp/[contents]"]
        _site/["/tmp/_site/"]
        quarto{{"quarto"}}
    end
  end
  pages/contents --> contents
  site-config.yml --- yq
  /static/contents --> contents
  /static/_quarto.yml --- yq
  yq -->|yq -i '...' _quarto.yml | _quarto.yml
  _quarto.yml --> quarto
  contents --> quarto

>   quarto -->|quarto render /tmp/| _site/ --> out/

Reviewer 2: Change `out/` to `data/` here if applicable (i.e., I'm correct in the above comment)

## How the image is built

The images is built using GitHub actions and [deployed to GitHub's container registry](https://github.com/hubverse-org/hub-dash-site-builder/pkgs/container/hub-dash-site-builder/406574245?tag=latest).

We initially created the build workflow by referencing [GitHub's Publishing Docker images
guide](https://docs.github.com/en/actions/use-cases-and-examples/publishing-packages/publishing-docker-images#publishing-images-to-github-packages),
but we also wanted to be able to test the image and only publish when we
created a tag or release AND we wanted to be mindful of good security practices
with GitHub workflows (especially with respect to the principle of least
permission).

The workflow that we came up with is called [build-container.yaml](https://github.com/hubverse-org/hub-dash-site-builder/blob/main/.github/workflows/build-container.yaml) and it contains three jobs:

### Build image

- **purpose**: build and test the docker image and then save it as an artifact
- **permissions**: read-all
- **runs on**: pull request, push to main, tags that start with `v`, and manual
  trigger.

There is a bit of a dance that's required for this job to run, which is

> described in GitHub's Publishing Docker images guide (see above for link).

Reviewer 2: Can we just link this page again here?

Author: good point! I think I was originally trying to [avoid using duplicate links to the same page](https://www.nngroup.com/articles/duplicate-links/), but you are correct that in this context it makes sense, especially given this guideline from cal tech's [digital accessibility guide on links](https://dap.berkeley.edu/learn/concepts/links)

> If you link to the same thing twice on a page, use the same link text in both places.

```suggestion
described in [GitHub's Publishing Docker images
guide](https://docs.github.com/en/actions/use-cases-and-examples/publishing-packages/publishing-docker-images#publishing-images-to-github-packages).
```
